pubsub/gcp/pubsub/pubsub.go

```diff
@@ -235,20 +266,76 @@ func (g *GCPPubSub) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handle
 }
 
 func (g *GCPPubSub) handleSubscriptionMessages(topic *gcppubsub.Topic, sub *gcppubsub.Subscription, handler pubsub.Handler) error {
-	err := sub.Receive(context.Background(), func(ctx context.Context, m *gcppubsub.Message) {
-		msg := &pubsub.NewMessage{
-			Data:  m.Data,
-			Topic: topic.ID(),
-		}
+	// Limit the number of attempted reconnects we make.
+	reconnAttempts := make(chan struct{}, g.metadata.MaxReconnectionAttempts)
+	for i := 0; i < g.metadata.MaxReconnectionAttempts; i++ {
+		reconnAttempts <- struct{}{}
+	}
 
-		err := handler(ctx, msg)
+	readReconnectAttemptsRemaining := func() int { return len(reconnAttempts) }
+
+	// Periodically refill the reconnect attempts channel to avoid
+	// exhausting all the refill attempts due to intermittent issues
+	// occurring over a longer period of time.
+	reconnCtx, reconnCancel := context.WithCancel(g.ctx)
+	defer reconnCancel()
+
+	go func() {
+		// Calculate refill interval relative to MaxReconnectionAttempts and ConnectionRecoveryInSec
+		// in order to allow reconnection attempts to be exhausted in case of a permanent issue (e.g. wrong subscription name).
+		refillInterval := time.Duration(2*g.metadata.MaxReconnectionAttempts*g.metadata.ConnectionRecoveryInSec) * time.Second
```

Comment:
Is this an implementation of a token bucket? If so, can we reuse an existing library that does that?

Reply:
it might make sense to reuse existing dapr retry tools if they match the use case, but does that mean that the very similar retry loop implementation in the Azure Service Bus pub/sub component here https://github.com/dapr/components-contrib/blob/master/pubsub/azure/servicebus/servicebus.go#L384 also has to change?

Reply:
I have reviewed the code for 'dapr/kit/retry' available here https://github.com/dapr/kit/blob/main/retry/retry.go and the backoff library it uses (https://pkg.go.dev/github.com/cenkalti/backoff) and I don't think it's directly applicable since it has different behavior - it would NOT refill the number of reconnection attempts after a period of no errors, which means that it would be prone to stop reconnecting if the reconnect attempts are slowly exhausted over a long period of time. I am therefore resolving this without change.